Kafka EventBus

dfarr opened the pull request:
Implement Kafka EventBus

dfarr committed:
Initial scaffolding for kafka eventbus

Signed-off-by: David Farr <david_farr@intuit.com>

premadk committed:
Add kafka eventbus spec (#12)

Signed-off-by: Prema devi Kuppuswamy <premadk@gmail.com>
Signed-off-by: David Farr <david_farr@intuit.com>

bilalba committed:
in-cluster kafka (#14)

* in-cluster kafka

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* fix amd64

* add kafka clusters through kustomization

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* revert arm64 change

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* addressing comments

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* Consolidate manifests

Signed-off-by: David Farr <david_farr@intuit.com>

* Fix indentation

Signed-off-by: David Farr <david_farr@intuit.com>

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>
Signed-off-by: David Farr <david_farr@intuit.com>
Co-authored-by: David Farr <david_farr@intuit.com>
Signed-off-by: David Farr <david_farr@intuit.com>

bilalba committed:
Kafka Validations (#16)

* webhook validations

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* add Exotic Kafka installer

* Change order of EventBus reconcilation

Calling client.Status().Update() before calling client.Update()
persists the status for exotic evnetbuses, otherwise the status
is not perisisted.

Signed-off-by: David Farr <david_farr@intuit.com>

* requeue reconciler

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

* Make additional copy of obj to update eb status

Signed-off-by: David Farr <david_farr@intuit.com>

* oxford comma

Co-authored-by: David Farr <david_farr@intuit.com>

* stylistic change

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>

Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>
Signed-off-by: David Farr <david_farr@intuit.com>
Co-authored-by: David Farr <david_farr@intuit.com>
Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Implement kafka eventbus eventsource (#11)

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Enable kafka functional tests

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Kafka eventbus sensor (#18)

Initial implementation of kafka eventbus sensor

---------

Signed-off-by: David Farr <david_farr@intuit.com>
Signed-off-by: Bilal Bakht Ahmad <tringingly@gmail.com>
Co-authored-by: Bilal Bakht Ahmad <tringingly@gmail.com>
Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Enable sasl support for kafka eventbus

TODO: consolidate sarama config initialization into common code
shared between eventbus eventsource and sensor

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add offset manager

Enables the ability to ignore messages after pod failure to
ensure correct resiliency.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Enable batching

A kafka transaction is expensive, to improve throughput we can
batch together incoming messages and include all the resulting
output information in a single transaction.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add argo-events-sa service account to all e2e tests

This service account has authorization to use leases, which is
required for kubernetes based leader election.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Consolidate kafka eventbus config

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Action function need not return an error

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Enable condition resets

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Change event time check logic

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Bypass leader election for kafka eventbus sensor

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Guard against batch with no messages

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Use message timestamp for resets instead of event timestamp

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add version to kafka config

Signed-off-by: David Farr <david_farr@intuit.com>

premadk committed:
fix version in config (#19)

Signed-off-by: Prema devi Kuppuswamy <premadk@gmail.com>
Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Remove kafka streamconfig

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Attach access secret for kafka eventbus and sensor

Signed-off-by: David Farr <david_farr@intuit.com>

AalokAhluwalia commented on controllers/eventbus/controller.go:
```diff
@@ -57,14 +57,20 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		log.Errorw("reconcile error", zap.Error(reconcileErr))
 	}
+
+	// client.Update() mutates busCopy so we need to make an
```
Why is the update mutating the busCopy? In the other fork at develop/controllers/sensormetadata/controller.go#L175-L183, we don't need a copy when saving the status. What other fields is Update changing other the status?

dfarr replied:
In the call to `r.client.Update(ctx, busCopy)`, the `busCopy` variable is a pointer and can be mutated by the Update function, resulting in an unexpected object being passed to `r.client.Status().Update(ctx, busCopy)`. This is a bug we discovered while doing this PR.

All this is probably moot now since there is a [separate PR](https://github.com/argoproj/argo-events/pull/2466) to address this bug.

AalokAhluwalia commented on eventbus/driver.go:
```diff
@@ -35,12 +37,12 @@ func GetEventSourceDriver(ctx context.Context, eventBusConfig eventbusv1alpha1.B
 
 	var eventBusType apicommon.EventBusType
 	switch {
-	case eventBusConfig.NATS != nil && eventBusConfig.JetStream != nil:
```
Might be a good check to keep in to make sure multiple specs aren't defined.

dfarr replied:
But permutations increase exponentially, so it quickly gets out of control
```
1. eventBusConfig.NATS != nil && eventBusConfig.JetStream != nil
2. eventBusConfig.NATS != nil && eventBusConfig.KafkaStream != nil
3. eventBusConfig.JetStream != nil && eventBusConfig.KafkaStream != nil
4. eventBusConfig.NATS != nil && eventBusConfig.JetStream != nil && eventBusConfig.KafkaStream != nil
```

Yuck!

AalokAhluwalia commented on eventbus/driver.go:
```diff
@@ -79,12 +83,12 @@ func GetSensorDriver(ctx context.Context, eventBusConfig eventbusv1alpha1.BusCon
 
 	var eventBusType apicommon.EventBusType
 	switch {
-	case eventBusConfig.NATS != nil && eventBusConfig.JetStream != nil:
```
Same comment as above

AalokAhluwalia commented on eventbus/kafka/base/kafka.go:
```diff
@@ -0,0 +1,112 @@
+	// consumer config
+	config.Consumer.IsolationLevel = sarama.ReadCommitted
+	config.Consumer.Offsets.AutoCommit.Enable = false
+
```
Shouldn't this Config method also set the `ConsumerGroup.GroupName`?

dfarr replied:
GroupName is specified on the consumer, not the config. You can see this [here](https://github.com/intuit-data-os/argo-events/pull/10/files/690d42a83591d487dd4b2359f0f85b5444dd7d9f#diff-b778225ecfb035ee9a44077f3f987fd20bc3381f1786485f86639419b0c9196fR118)

AalokAhluwalia commented on common/leaderelection/leaderelection.go:
```diff
@@ -40,7 +40,7 @@ type LeaderCallbacks struct {
 func NewElector(ctx context.Context, eventBusConfig eventbusv1alpha1.BusConfig, clusterName string, clusterSize int, namespace string, leasename string, hostname string) (Elector, error) {
 	switch {
-	case strings.ToLower(os.Getenv(common.EnvVarLeaderElection)) == "k8s":
+	case eventBusConfig.Kafka != nil || strings.ToLower(os.Getenv(common.EnvVarLeaderElection)) == "k8s":
```
In sensors/listener.go you mention kafka doesn't require leader election? If sources don't do leader election also, maybe return an error mentioning that in this function

dfarr replied:
Eventsources still need a leader election (under some circumstances, it's a little confusing). So unfortunately we still need this for the event source.

dfarr replied:
[Here's the logic](https://github.com/intuit-data-os/argo-events/blob/master/eventsources/eventing.go#L376-L393) that determines if the eventsource needs a leader election or not. The important thing to note is there are circumstances where an eventsource with a kafka eventbus requires a leader election.

AalokAhluwalia commented on eventbus/kafka/eventsource/source_conn.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func (c *KafkaSourceConnection) Publish(ctx context.Context, msg common.Message) error {
+	key := base.EventKey(msg.EventSourceName, msg.EventName)
```
Shouldn't the key also contain information about the source's namespace? Is there any chance for error if two sources in different namespaces have the same event name / topic also

dfarr replied:
Hmm... the eventbus is a namespace scoped object, so on first thought including namespace in the key seems unnecessary, but it would be possible to create two eventbuses and point them at the same topic. Let me see if we have access to the namespace name here...

dfarr replied:
I'm going to leave this as is. If two eventsources had the same name across two different namespaces **and** used the same topic then there would be much bigger fish to fry - sensors could potentially consume messages from the eventsource in another namespace if the event names were also the same.

What we really should do is guard against this altogether. If an eventsource already exists that is wired up to topic "x" and someone else creates another eventbus wired up to topic "x" we should probably reject creation of the second eventbus. I feel this can probably be a followup PR.

AalokAhluwalia commented on eventbus/kafka/eventsource/source_conn.go:
```diff
@@ -0,0 +1,46 @@
+	})
+
+	if err != nil {
+		return err
```
Maybe log an error here?

dfarr replied:
I think we would just be duplicating the information logged [here](https://github.com/intuit-data-os/argo-events/blob/feature/kafka-eventbus/eventsources/eventing.go#L553-L554)

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_transaction.go:
```diff
@@ -0,0 +1,94 @@
+func (t *KafkaTransaction) Commit(session sarama.ConsumerGroupSession, messages []*sarama.ProducerMessage, offset int64, metadata string) error {
+	// No need for a transaction if no messages
+	if len(messages) == 0 {
+		session.MarkOffset(t.Topic, t.Partition, offset, metadata)
```
If messages is empty, what would the offset be? is there any need to commit this?

dfarr replied:
Good catch! It is still very important to commit when there are no messages, this can happen for two reasons.

1) Because the action topic is the final topic, we always bump the offset without producing any messages.
2) The trigger topic can bump the offset without producing a message. Imagine you have have a trigger with  condition `a && b` and consume a stream of `a`'s. Every time we consume a new `a` we want to bump the offset, but we can't produce a message because the trigger condition is not yet satisfied.

AalokAhluwalia commented on pkg/apis/eventbus/v1alpha1/kafka_eventbus.go:
```diff
@@ -0,0 +1,46 @@
+
+type KafkaConsumerGroup struct {
+	// The name for the consumer group to use
+	GroupName string `json:"groupName,omitempty" protobuf:"bytes,1,opt,name=groupName"`
```
Do you need to add the +optional comment here

dfarr replied:
Good catch, I'll update. This is definitely not needed because we set a [default value](https://github.com/intuit-data-os/argo-events/pull/10/files/690d42a83591d487dd4b2359f0f85b5444dd7d9f#diff-b778225ecfb035ee9a44077f3f987fd20bc3381f1786485f86639419b0c9196fR48-R53)

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_sensor.go:
```diff
@@ -0,0 +1,355 @@
+func NewKafkaSensor(kafkaConfig *eventbusv1alpha1.KafkaConfig, sensor *sensorv1alpha1.Sensor, hostname string, logger *zap.SugaredLogger) *KafkaSensor {
+	topics := &Topics{
+		event:   kafkaConfig.Topic,
+		trigger: fmt.Sprintf("%s.%s.%s", kafkaConfig.Topic, sensor.Name, "trigger"),
```
For the trigger topics, same comment as before, should it also include the namespace to include the possibility that  there are two sensors in different namespaces with the same name.

dfarr replied:
Yes definitely! The namespace is already included in the topic name itself [here](https://github.com/intuit-data-os/argo-events/pull/10/files/690d42a83591d487dd4b2359f0f85b5444dd7d9f#diff-88e14955e32e22bdfa25772058dc7148da8a6fbc42b6ac426c061c80ef846fcfR33)

dfarr replied:
But I think we should think through the option of the user specifying the topic names manually - do you think users may want to provide the trigger and action topic name manually too?

dfarr replied:
Personally I feel it's sufficient to (optionally) have the user specify the eventbus {event} topic and then derive the {trigger, action} topic names from the name of this topic (as I do currently)

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_sensor.go:
```diff
@@ -0,0 +1,355 @@
+	}
+
+	if _, ok := s.triggers[triggerName]; !ok {
+		expr, err := govaluate.NewEvaluableExpression(strings.ReplaceAll(depExpression, "-", "\\-"))
```
What is line 165 `NewEvaluableExpression(strings.ReplaceAll` for ? Does the argo events code do this in other places?

dfarr replied:
This creates a new evaluatable expression from the string that's provided in the trigger's condition field. It's the same library as used today in jetstream [here](https://github.com/intuit-data-os/argo-events/blob/master/eventbus/jetstream/sensor/trigger_conn.go#L68).

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_sensor.go:
```diff
@@ -0,0 +1,355 @@
+	defer s.Disconnect()
+
+	for {
+		if len(s.triggers) != len(s.sensor.Spec.Triggers) || !s.triggers.Ready() {
```
Is it possible that the triggers would never be ready and this would loop forever. Would it be helpful to add a max wait duration?

dfarr replied:
Doesn't hurt, I can add one!

dfarr replied:
So I implemented this and then I realized, even if we waited a max duration and return, the function would just be re-invoked in `sensors/listener.go` anyways. Furthermore, I realized that Connect and Subscribe are guaranteed to be called for each trigger (and if this happens then the if condition will be satisfied) so I don't think this is actually necessary.

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_sensor.go:
```diff
@@ -0,0 +1,355 @@
+		}
+
+		if !trigger.Filter(trigger.depName, event) {
+			continue
```
Maybe add a debug log that message was filtered for X trigger

AalokAhluwalia commented on eventbus/kafka/sensor/kafka_sensor.go:
```diff
@@ -0,0 +1,355 @@
+	return messages, offset, nil
+}
+
+func (s *KafkaSensor) Action(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
```
Looks like for event/action/trigger the second return value is always msg.offset + 1. Maybe remove this as a return value if it's always the same

dfarr replied:
It's not always the same, the trigger is the odd one out, it is set to the smallest required offset across all triggers. That value is set [here](https://github.com/intuit-data-os/argo-events/pull/10/files/690d42a83591d487dd4b2359f0f85b5444dd7d9f#diff-b778225ecfb035ee9a44077f3f987fd20bc3381f1786485f86639419b0c9196fR336)

AalokAhluwalia commented:
And then just one genreal question for the above triggering logic. Suppose you have the condition (a&&b) || (b&&c) and event b has already come in so the expression looks like (a && true) || (true && c). What will happen if in the next batch, messages a and c come. Will there be two triggers?

dfarr committed:
Attach secrets to eventsource and sensor when using kafka eventbus

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr commented:
> And then just one genreal question for the above triggering logic. Suppose you have the condition (a&&b) || (b&&c) and event b has already come in so the expression looks like (a && true) || (true && c). What will happen if in the next batch, messages a and c come. Will there be two triggers?

The ordering of messages in a batch is guaranteed (thanks kafka) so the `a` event would satisfy the trigger (a message to the action topic would be published), the triggers state would be reset, and then the first message added to the fresh state of the trigger would be the `c` event.

So stepping through this (the number preceding the event represents the offset)
```
batch 1:
0: b

state:
(a && true) || (true && c)

transaction:
offset: 0
messages: []

batch 2:
1: a
2: c

state:
(a && b) || (b && true)

transaction:
offset: 2
messages: [action]
```

dfarr committed:
Remove accessSecret from kafka config and make group name optional

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Log when event is filtered out

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add batch tests

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Provide helpful message if topic(s) do not exist and fail fast

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add timing asserts to batch tests

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Nullify batch to reset

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Batch every second

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Revert eventbus controller status copy

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Update comments

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Revert ci pull_request branches

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Use dashes ('-') instead of dots ('.') in default kafka names

Creating a topic with a dot gives the following warning
WARNING: Due to limitations in metric names, topics with a period ('.') or underscore ('_') could collide. To avoid issues it is best to use either, but not both.

Avoiding this altogether by switching to dashes.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add comments

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Move kafka specs to test-functional only

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add batch comment

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Remove exotic from kafka eventbus spec

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add additional controller tests for eventsource and sensor

Test against all eventbus types (nats, jetstream, kafka) and ensure
that kafka secrets are attached.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add Kafka EventBus docs

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Remove key from kafka trigger message

Signed-off-by: David Farr <david_farr@intuit.com>
